### ISSUE TITLE
Long queue time on horovod timeline

### ISSUE THREAD

**Issue by KimMeen** [horovod/horovod#992]
**Environment:**
1. Framework: Keras(Tensorflow Backend)
2. Framework version: Tensorflow-gpu 1.8.0
3. Horovod version: 0.16.1
4. MPI version: 4.0.0
5. CUDA version: 9.0
6. NCCL version: 2.4.2
7. Python version: 3.6
8. OS and version: Ubuntu 16.04

**My question:**
Currently I am using `tensorflow_synthetic_benchmark.py` to test the distributed environment, but I found that the running time and processing efficiency were much lower than in a single node when I tried to distributed the code between two nodes. 
Some environment details are listed below:
```
Node1(tpc1):  Intel(R) Xeon(R) CPU E3-1220 v5 @ 3.00GHz with 1050 Ti
Node2(tpc2):  Intel(R) Core(TM) i7-8700K CPU @ 3.70GHz with 1060
Network: 95Mbit/s LAN calculated by iperf (I know this might be one of the bottlenecks)

Here are some network testing details:
tpc1@tpc1-outotec:~$ iperf -u -c 192.168.1.1 -b 1000M -t 60 -i 10
------------------------------------------------------------
Client connecting to 192.168.1.1, UDP port 5001
Sending 1470 byte datagrams
UDP buffer size:  208 KByte (default)
------------------------------------------------------------
[  3] local 192.168.1.12 port 41939 connected with 192.168.1.1 port 5001
[ ID] Interval       Transfer     Bandwidth
[  3]  0.0-10.0 sec   113 MBytes  95.1 Mbits/sec
[  3] 10.0-20.0 sec   114 MBytes  95.2 Mbits/sec
[  3] 20.0-30.0 sec   113 MBytes  95.2 Mbits/sec
[  3] 30.0-40.0 sec   113 MBytes  95.2 Mbits/sec
[  3] 40.0-50.0 sec   113 MBytes  95.2 Mbits/sec
[  3] 50.0-60.0 sec   113 MBytes  95.2 Mbits/sec
[  3]  0.0-60.0 sec   681 MBytes  95.2 Mbits/sec
[  3] Sent 485676 datagrams
[  3] Server Report:
[  3]  0.0-60.0 sec   681 MBytes  95.2 Mbits/sec   0.484 ms    0/485675 (0%)
[  3]  0.0-60.0 sec  1 datagrams received out-of-order

```
If I run the code on tpc1 or tpc2 solely, the performance is around 44 images/sec and 76 images/sec(and pretty fast), but if I run on both of them, the rate will drop to 1.7 images/sec(and very slow).

When I check the timeline, I found that the queue time seems the reason that slows down the training process. According to [this](https://github.com/horovod/horovod/blob/master/docs/timeline.md), queue happens when reduction is done with NCCL, and the previous NCCL operation did not finish yet. So what the reason of that in my case?

Here is my command to run the test on two nodes:

`mpirun --allow-run-as-root --tag-output -np 2 -H localhost:1,tpc2:1 -bind-to none -map-by slot -mca pml ob1 -mca btl ^openib -mca btl_tcp_if_exclude lo -x NCCL_DEBUG=INFO -x NCCL_MIN_NRINGS=1 -x NCCL_SOCKET_IFNAME=eno1,enp0s31f6 -x HOROVOD_TIMELINE='/mnt/data/project/Linux Distributed Version/imagenet_example/timeline.json' -x HOROVOD_TIMELINE_MARK_CYCLES=1 -x XDG_VTNR -x XDG_SESSION_ID -x XDG_GREETER_DATA_DIR -x CLUTTER_IM_MODULE -x SESSION -x GPG_AGENT_INFO -x TERM -x VTE_VERSION -x XDG_MENU_PREFIX -x SHELL -x CONDA_SHLVL -x QT_LINUX_ACCESSIBILITY_ALWAYS_ON -x CONDA_PROMPT_MODIFIER -x WINDOWID -x OLDPWD -x UPSTART_SESSION -x GNOME_KEYRING_CONTROL -x GTK_MODULES -x USER -x LS_COLORS -x QT_ACCESSIBILITY -x LD_LIBRARY_PATH -x CONDA_EXE -x UNITY_HAS_3D_SUPPORT -x XDG_SESSION_PATH -x XDG_SEAT_PATH -x SSH_AUTH_SOCK -x DEFAULTS_PATH -x XDG_CONFIG_DIRS -x UNITY_DEFAULT_PROFILE -x DESKTOP_SESSION -x PATH -x QT_IM_MODULE -x QT_QPA_PLATFORMTHEME -x CONDA_PREFIX -x XDG_SESSION_TYPE -x PWD -x JOB -x XMODIFIERS -x GNOME_KEYRING_PID -x LANG -x GDM_LANG -x MANDATORY_PATH -x COMPIZ_CONFIG_PROFILE -x IM_CONFIG_PHASE -x GDMSESSION -x SESSIONTYPE -x GTK2_MODULES -x SHLVL -x HOME -x XDG_SEAT -x LANGUAGE -x GNOME_DESKTOP_SESSION_ID -x UPSTART_INSTANCE -x CONDA_PYTHON_EXE -x UPSTART_EVENTS -x XDG_SESSION_DESKTOP -x LOGNAME -x COMPIZ_BIN_PATH -x DBUS_SESSION_BUS_ADDRESS -x XDG_DATA_DIRS -x QT4_IM_MODULE -x CONDA_DEFAULT_ENV -x LESSOPEN -x INSTANCE -x UPSTART_JOB -x XDG_RUNTIME_DIR -x DISPLAY -x XDG_CURRENT_DESKTOP -x GTK_IM_MODULE -x LESSCLOSE -x XAUTHORITY -x _ -x _HOROVOD_SECRET_KEY /mnt/environment/python '/mnt/data/project/Linux Distributed Version/imagenet_example/tensorflow_synthetic_benchmark.py' --model ResNet50 --batch-size 16`

And this was the NCCL information for the command above:

```
[1,0]<stdout>:tpc1-outotec:15322:15330 [0] NCCL INFO NET/Socket : Using [0]enp0s31f6:192.168.1.12<0>
[1,0]<stdout>:tpc1-outotec:15322:15330 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so).
[1,0]<stdout>:
[1,0]<stdout>:tpc1-outotec:15322:15330 [0] misc/ibvwrap.cu:63 NCCL WARN Failed to open libibverbs.so[.1]
[1,0]<stdout>:NCCL version 2.4.2+cuda9.0
[1,1]<stdout>:tpc2-outotec:17291:17306 [0] NCCL INFO NET/Socket : Using [0]eno1:192.168.1.1<0>
[1,1]<stdout>:tpc2-outotec:17291:17306 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so).
[1,1]<stdout>:
[1,1]<stdout>:tpc2-outotec:17291:17306 [0] misc/ibvwrap.cu:63 NCCL WARN Failed to open libibverbs.so[.1]
[1,0]<stdout>:tpc1-outotec:15322:15330 [0] NCCL INFO Setting affinity for GPU 0 to 0f
[1,0]<stdout>:tpc1-outotec:15322:15330 [0] NCCL INFO comm 0x7f55782eddc0 rank 0 nranks 2 cudaDev 0 nvmlDev 0
[1,1]<stdout>:tpc2-outotec:17291:17306 [0] NCCL INFO Setting affinity for GPU 0 to 0fff
[1,1]<stdout>:tpc2-outotec:17291:17306 [0] NCCL INFO comm 0x7f9de82e9a70 rank 1 nranks 2 cudaDev 0 nvmlDev 0
[1,1]<stdout>:tpc2-outotec:17291:17306 [0] NCCL INFO CUDA Dev 0[0], Socket NIC distance :  PHB
[1,0]<stdout>:tpc1-outotec:15322:15330 [0] NCCL INFO CUDA Dev 0[0], Socket NIC distance :  PHB
[1,0]<stdout>:tpc1-outotec:15322:15330 [0] NCCL INFO NCCL_MIN_NRINGS set by environment to 1.
[1,1]<stdout>:tpc2-outotec:17291:17306 [0] NCCL INFO NCCL_MIN_NRINGS set by environment to 1.
[1,0]<stdout>:tpc1-outotec:15322:15330 [0] NCCL INFO Channel 00 :    0   1
[1,1]<stdout>:tpc2-outotec:17291:17306 [0] NCCL INFO Ring 00 : 0 -> 1 [receive] via NET/Socket/0
[1,0]<stdout>:tpc1-outotec:15322:15330 [0] NCCL INFO Ring 00 : 1 -> 0 [receive] via NET/Socket/0
[1,1]<stdout>:tpc2-outotec:17291:17306 [0] NCCL INFO Ring 00 : 1 -> 0 [send] via NET/Socket/0
[1,0]<stdout>:tpc1-outotec:15322:15330 [0] NCCL INFO Ring 00 : 0 -> 1 [send] via NET/Socket/0
[1,0]<stdout>:tpc1-outotec:15322:15330 [0] NCCL INFO Using 256 threads, Min Comp Cap 6, Trees disabled
[1,0]<stdout>:tpc1-outotec:15322:15330 [0] NCCL INFO comm 0x7f55782eddc0 rank 0 nranks 2 cudaDev 0 nvmlDev 0 - Init COMPLETE
[1,0]<stdout>:tpc1-outotec:15322:15330 [0] NCCL INFO Launch mode Parallel
[1,1]<stdout>:tpc2-outotec:17291:17306 [0] NCCL INFO comm 0x7f9de82e9a70 rank 1 nranks 2 cudaDev 0 nvmlDev 0 - Init COMPLETE
[1,0]<stdout>:Running benchmark...
[1,0]<stdout>:Iter #0: 1.7 img/sec per GPU
[1,0]<stdout>:Iter #1: 1.7 img/sec per GPU
[1,0]<stdout>:Iter #2: 1.7 img/sec per GPU
[1,0]<stdout>:Iter #3: 1.7 img/sec per GPU
[1,0]<stdout>:Iter #4: 1.7 img/sec per GPU
[1,0]<stdout>:Iter #5: 1.7 img/sec per GPU
[1,0]<stdout>:Iter #6: 1.7 img/sec per GPU
[1,0]<stdout>:Iter #7: 1.7 img/sec per GPU
[1,0]<stdout>:Iter #8: 1.7 img/sec per GPU
[1,0]<stdout>:Iter #9: 1.7 img/sec per GPU
[1,0]<stdout>:Img/sec per GPU: 1.7 +-0.0
[1,0]<stdout>:Total img/sec on 2 GPU(s): 3.5 +-0.0
```
This is the timeline screenshot for this running:
![Screenshot01](https://user-images.githubusercontent.com/33118451/55604499-d4ab2e80-57bb-11e9-842e-5e3ec01cce05.png)

I have several assumptions but I am not sure which one contributes to this problem:
1. Network limitation
2. ResNet50 is still not complex enough to see the scaling (high communication overhead in this case)
3. Two nodes are too few to see the scaling

Any ideas to solve this problem? Thank you!

**Comment by alsrgv** [comment]
@KimMeen, unfortunately, 100Mbit NIC will not cut it at all at this point.

You can try to add fp16 compression (`hvd.DistributedOptimizer(opt, compression=hvd.Compression.fp16)`), but that may double your img/sec at best.  We're looking into techniques for deeper gradient compression, but they have not materialized yet.

Can you get at least 10Gbit NIC?

**Comment by KimMeen** [comment]
@alsrgv, I tried fp 16 compression but that couldn't help me a lots in my current network condition, and I am not sure which factors contributed to this relatively low bandwidth(NIC, router or cables). 

So now I have two options here, the one is to improve the bandwidth to at least 10Gbit, and another one is to install these two graphic cards in a single machine and then launch the task by using this command: 

`horovodrun --verbose -np 2 -H localhost:2 python training.py`

Does the second option seems more reasonable for me? Thank you :D

**Comment by alsrgv** [comment]
@KimMeen, definitely, if you can install both cards in the same machine you'll have much better performance.

**Comment by KimMeen** [comment]
Yes indeed, I tried this method and there was no queue anymore in my timeline. Thank you for the help! @alsrgv

Now I close this issue.
